### PR TITLE
(RE-12205) Add retry + other time servers to msi signing

### PR DESCRIPTION
**Pull request by denis-susan**
This pr solves latency issues of timeservers by:
* adding multiple time servers
* adding a retry mechanism of 5 tries per server.

### CODE REVIEW

**Comment by melissa** [comment]
@denis-susan thanks for submitting this! Any reason you're only adding the retries to the sha1 signature and not the sha256 signature?

File: lib/packaging/sign/msi.rb [hunk 1]
```diff
@@ -66,14 +66,24 @@ def sign(target_dir = 'pkg')
...
+                    -h sha1 \
+                    -pkcs12 "#{Pkg::Config.msi_signing_cert}" \
+                    -pass "#{Pkg::Config.msi_signing_cert_pw}" \
+                    -t "$timeserver" \
+                    -in "C:/#{work_dir}/$msi" \
+                    -out "C:/#{work_dir}/signed-$msi)
+          if [ $ret == "Succeeded" ]; then break; fi
+          done;
+          if [ $ret == "Succeeded" ]; then break; fi
+          done;
```
Comment (mwaggett) [inline lib/packaging/sign/msi.rb]: I assume this duplication was a copy+paste mistake?

**Comment by mwaggett** [comment]
seconding @melissa: since timeouts appear to happen for both signature types, I think we should probably add retries to both, unless you know of a reason not to.

**Comment by mwaggett** [comment]
also @denis-susan could you update the commit message to include the relevant ticket title? https://tickets.puppetlabs.com/browse/RE-12205

File: lib/packaging/sign/msi.rb [hunk 2]
```diff
@@ -66,14 +66,24 @@ def sign(target_dir = 'pkg')
...
-          -t "http://timestamp.verisign.com/scripts/timstamp.dll" \
-          -in "C:/#{work_dir}/$msi" \
-          -out "C:/#{work_dir}/signed-$msi"
+        timeservers=(http://timestamp.verisign.com/scripts/timstamp.dll
+          http://timestamp.globalsign.com/scripts/timstamp.dll
+          http://www.startssl.com/timestamp)
+          tries=5
+          for timeserver in "${timeservers[@]}"; do
+          for ((try=1; try<=$tries; try++)) do
+          ret=$(/cygdrive/c/tools/osslsigncode-fork/osslsigncode.exe" sign \
```
Comment (mwaggett) [inline lib/packaging/sign/msi.rb]: I think you've got an errant `"` here

**Comment by mwaggett** [comment]
Has this been tested? I created https://confluence.puppetlabs.com/display/SRE/How+to+test+%28certain%29+changes+to+packaging to explain how to test changes to packaging. 
For this, you'll probably want to build a new windows package (`bundle exec build puppet-agent windows2012r2-x64`) and then run `bundle exec sign` to test these changes. 
I *think* you should be able to run that without a problem since you're in the 'release' group and windows signing doesn't require a passphrase (this is why you need to build new packages - if you retrieved any rpms or debs, you'd be prompted for a passphrase when trying to sign those), but if you run into trouble, we can test it out for you.

File: lib/packaging/sign/msi.rb (hunk 1 above)
Comment (denis-susan) [reply lib/packaging/sign/msi.rb]: No, this is to break the other loop (there's 2 loops, one to retry the current server and another to try different servers, that loop just breaks out of the loops on success

**Comment by denis-susan** [comment]
> @denis-susan thanks for submitting this! Any reason you're only adding the retries to the sha1 signature and not the sha256 signature?

currently only the sha1 timestamp server has issues. If we need to add the same mechanism for sha256 it should be trivial.

**Comment by denis-susan** [comment]
> Has this been tested? I created https://confluence.puppetlabs.com/display/SRE/How+to+test+%28certain%29+changes+to+packaging to explain how to test changes to packaging.
> For this, you'll probably want to build a new windows package (`bundle exec build puppet-agent windows2012r2-x64`) and then run `bundle exec sign` to test these changes.
> I _think_ you should be able to run that without a problem since you're in the 'release' group and windows signing doesn't require a passphrase (this is why you need to build new packages - if you retrieved any rpms or debs, you'd be prompted for a passphrase when trying to sign those), but if you run into trouble, we can test it out for you.

I cant test, it asks me for a password to mozart

File: lib/packaging/sign/msi.rb (hunk 1 above)
Comment (mwaggett) [reply lib/packaging/sign/msi.rb]: ohhhh, I see. the lack of indentation threw me off.

**Comment by mwaggett** [comment]
gahh, @denis-susan I forgot that since this is just signing, you can test with the vanagon generic builder, since we sign at build-time now.
Looks like syntax is off somewhere: https://jenkins-master-prod-1.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/337/

**Comment by denis-susan** [comment]
Passing build: https://jenkins-master-prod-1.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/339/

**Pull request merged**